File: processor/src/main/java/com/qualityminds/lazyval/processor/internal/codegen/JpaGenerator.java
package com.qualityminds.lazyval.processor.internal.codegen;

import com.palantir.javapoet.*;
import com.qualityminds.lazyval.collections.NonEmptySet;
import com.qualityminds.lazyval.processor.spi.Generator;
import com.qualityminds.lazyval.processor.spi.GeneratorResult;
import com.qualityminds.lazyval.processor.spi.ValidatedGeneratorElement;

import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

/**
 * Generates a JPA {@code AttributeConverter} for each domain-primitive.
 *
 * <h1>Null invariants</h1>
 * Both {@code convertToDatabaseColumn} and {@code convertToEntityAttribute} guard against
 * {@code null} explicitly: a {@code null} column value maps to a {@code null} entity attribute,
 * and a {@code null} entity attribute maps to a {@code null} column value.
 * Java's type system does not enforce nullability; both directions are therefore transparent to
 * {@code null} at runtime, regardless of whether the factory method can return {@code null}.
 */
// must only be public for ServiceLoader, but it is not part of the API
// tag::docu[]
public class JpaGenerator implements Generator {

    private static final String GENERATOR_ID = "jpa";
    private static final String OPTION_GENERATED_PACKAGE = "lazyval.jpa.package";

    private static final AnnotationSpec CONVERTER_ANNOTATION = AnnotationSpec.builder(ClassName.get("jakarta.persistence", "Converter"))
            .addMember("autoApply", "$L", "true")
            .build();

    @Override
    public String generatorId() {
        return GENERATOR_ID;
    }

    @Override
    public Collection<String> requiredClasspath() {
        return List.of("jakarta.persistence.AttributeConverter");
    }

    @Override
    public Set<String> supportedOptions() {
        return Set.of(OPTION_GENERATED_PACKAGE);
    }

    @Override
    public Stream<GeneratorResult> generate(NonEmptySet<ValidatedGeneratorElement> elements, Context context){
        // end::docu[]

        final String jpaConverterPackage = context.generatorPackage(OPTION_GENERATED_PACKAGE, "boundary.persistence.jpa");

        return elements.stream().map(JpaGenerator::buildAttributeConverter)
                .map(attributeConverterSpec -> JavaFile.builder(jpaConverterPackage, attributeConverterSpec)
                        .skipJavaLangImports(true)
                        .build())
                .map(javaFile -> new GeneratorResult.Java(new GeneratorResult.Metadata(
                        javaFile.packageName(), javaFile.typeSpec().name()),
                        javaFile.toString()));
    }

    private static TypeSpec buildAttributeConverter(ValidatedGeneratorElement validElement) {
        TypeMirror type = validElement.element().asType();
        var wrappedType = validElement.wrappedType();
        TypeName wrappedTypeName;
        if (wrappedType.isPrimitive()) {
            // Box primitive types for JPA generics
            wrappedTypeName = TypeName.get(wrappedType.typeMirror()).box();
        } else {
            wrappedTypeName = TypeName.get(wrappedType.typeMirror());
        }

        return TypeSpec.classBuilder(validElement.typeName().name() + "AttributeConverter")
                .addAnnotation(GeneratedStamp.forGenerator(JpaGenerator.class))
                .addAnnotation(CONVERTER_ANNOTATION)
                .addSuperinterface(ParameterizedTypeName.get(
                        ClassName.get("jakarta.persistence", "AttributeConverter"),
                        TypeName.get(type),
                        wrappedTypeName))
                .addModifiers(Modifier.PUBLIC)
                .addMethod(buildConvertToDatabaseColumn(validElement, wrappedTypeName, type))
                .addMethod(buildConvertToEntityAttribute(validElement, type, wrappedTypeName))
                .build();
    }

    private static MethodSpec buildConvertToEntityAttribute(ValidatedGeneratorElement validElement, TypeMirror type, TypeName wrappedTypeName) {
        String parameterName = "dbValue";
        return MethodSpec.methodBuilder("convertToEntityAttribute")
                .addModifiers(Modifier.PUBLIC)
                .returns(TypeName.get(type))
                .addParameter(wrappedTypeName, parameterName)
                .beginControlFlow("if(%s == null)".formatted(parameterName))
                .addStatement("return null")
                .endControlFlow()
                .addStatement("return $L", validElement.objectCreation(parameterName))
                .build();
    }

    private static MethodSpec buildConvertToDatabaseColumn(ValidatedGeneratorElement validElement, TypeName wrappedTypeName, TypeMirror type) {
        String parameterName = "type";
        // Jpa Converter needs to check for null due to boxing primitive types
        return MethodSpec.methodBuilder("convertToDatabaseColumn")
                .addModifiers(Modifier.PUBLIC)
                .returns(wrappedTypeName)
                .addParameter(TypeName.get(type), parameterName)
                .beginControlFlow("if(%s == null)".formatted(parameterName))
                .addStatement("return null")
                .endControlFlow()
                .addStatement(String.format("return %s.%s", parameterName, validElement.accessor()))
                .build();
    }
}

File: ksp/src/main/kotlin/com/qualityminds/lazyval/ksp/internal/codegen/JpaGenerator.kt
package com.qualityminds.lazyval.ksp.internal.codegen

import com.qualityminds.lazyval.collections.NonEmptySet
import com.qualityminds.lazyval.ksp.internal.codegen.GeneratedStamp.addGeneratedAnnotation
import com.qualityminds.lazyval.ksp.spi.Generator
import com.qualityminds.lazyval.ksp.spi.GeneratorResult
import com.qualityminds.lazyval.ksp.spi.StockGeneratorIds
import com.qualityminds.lazyval.ksp.spi.ValidatedKspGeneratorElement
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import java.util.stream.Stream

/**
 * Generates a JPA `AttributeConverter` for each domain-primitive.
 *
 * ## Null invariants
 *
 * Both `convertToDatabaseColumn` and `convertToEntityAttribute` use safe-call operators
 * (`?.`) to propagate `null` transparently: a `null` column value maps to a `null` entity
 * attribute, and a `null` entity attribute maps to a `null` column value.
 *
 * The generated converter always accepts and returns nullable types (`DomainType?` /
 * `InnerType?`) because JPA may pass `null` for optional columns regardless of whether the
 * domain type's factory method can return `null`. No distinction is made between nullable
 * and non-nullable factories; null is simply passed through in both directions.
 */
// tag::docu[]
class JpaGenerator : Generator {

    companion object {
        private const val OPTION_GENERATED_PACKAGE = "lazyval.jpa.package"
    }

    override fun generatorId(): String = StockGeneratorIds.JPA

    override fun requiredClasspath(): Set<String> = setOf("jakarta.persistence.AttributeConverter")

    override fun supportedOptions(): Set<String> {
        return setOf(OPTION_GENERATED_PACKAGE)
    }

    override fun generate(
        validatedElements: NonEmptySet<ValidatedKspGeneratorElement>,
        context: Generator.Context
    ): Stream<GeneratorResult> {
        // end::docu[]

        val packageName = context.generatorPackage(OPTION_GENERATED_PACKAGE, "boundary.persistence.jpa")

        return validatedElements.stream()
            .map { buildAttributeConverter(context, it) }
            .map { FileSpec.builder(packageName, it.name!!).addType(it).build() }
            .map { GeneratorResult.Kotlin(
                GeneratorResult.Metadata(it.packageName, it.name),
                it.toString()) }
    }


    private fun buildAttributeConverter(context: Generator.Context, validatedElement: ValidatedKspGeneratorElement): TypeSpec {
        val element = validatedElement.element
        val lazyvalTypeName = element.toClassName()
        // The unwrapped payload: a value class has no runtime form of its own, so the column type is
        // whatever it erases to.
        val payloadTypeName = validatedElement.payloadType.toTypeName()

        val converterClassName = "${validatedElement.name.flatName()}AttributeConverter"

        val convertToDatabaseColumn = buildConvertToDatabaseColumn(lazyvalTypeName, payloadTypeName, validatedElement)
        val convertToEntityAttribute = buildConvertToEntityAttribute(lazyvalTypeName, payloadTypeName, validatedElement)

        return TypeSpec.classBuilder(converterClassName)
            .addGeneratedAnnotation(JpaGenerator::class, context)
            .addAnnotation(
                AnnotationSpec.builder(ClassName("jakarta.persistence", "Converter"))
                    .addMember("autoApply = true")
                    .build()
            )
            .addSuperinterface(
                ClassName("jakarta.persistence", "AttributeConverter")
                    .parameterizedBy(
                        lazyvalTypeName.copy(nullable = true),
                        payloadTypeName.copy(nullable = true)
                    )
            )
            .addFunction(convertToDatabaseColumn)
            .addFunction(convertToEntityAttribute)
            .build()
    }

    private fun buildConvertToDatabaseColumn(
        lazyvalTypeName: ClassName,
        payloadTypeName: TypeName,
        validatedElement: ValidatedKspGeneratorElement
    ): FunSpec {
        val convertToDatabaseColumn = FunSpec.builder("convertToDatabaseColumn")
            .addModifiers(KModifier.OVERRIDE)
            .returns(payloadTypeName.copy(nullable = true))
            .addParameter("type", lazyvalTypeName.copy(nullable = true))
            .apply {
                addStatement("return ${validatedElement.kotlin.readOrNull("type")}")
            }
            .build()
        return convertToDatabaseColumn
    }

    private fun buildConvertToEntityAttribute(
        lazyvalTypeName: ClassName,
        payloadTypeName: TypeName,
        validatedElement: ValidatedKspGeneratorElement
    ): FunSpec {
        val parameterName = "dbValue"
        // Build convertToEntityAttribute method
        val convertToEntityAttribute = FunSpec.builder("convertToEntityAttribute")
            .addModifiers(KModifier.OVERRIDE)
            .returns(lazyvalTypeName.copy(nullable = true))
            .addParameter(parameterName, payloadTypeName.copy(nullable = true))
            .apply {
                addStatement("return ${validatedElement.kotlin.createOrNull(parameterName)}")
            }
            .build()
        return convertToEntityAttribute
    }
}